### ISSUE TITLE
[@typescript-eslint/explicit-function-return-type] Inconsistent on arrow functions

### ISSUE THREAD

**Issue by localnet** [typescript-eslint/typescript-eslint#1009]
**Repro**

```JSON
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "tsconfigRootDir": ".",
    "project": "tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended"
  ],
  "env": {
    "node": true
  },
  "rules": {
    "@typescript-eslint/promise-function-async": "error"
  }
}
```

```TS
public async listen(): Promise<Server> {
    if (this.server.listening) {
        return this.server;
    }

    return new Promise(resolve => { // 1) ERROR
        this.server.once("listening", () => resolve(this.server)); // 2) WORK AS EXPECTED
        this.server.listen(Number(process.env.SERVER_PORT), process.env.SERVER_HOST);
    });
}
```

**Expected Result**

No errors.

**Actual Result**

Error:
> 20:28  warning  Missing return type on function  @typescript-eslint/explicit-function-return-type

**Additional Info**

The line commented with `// 1) ERROR` cause the error, while the line commented with `// 2) WORK AS EXPECTED` work as expected. It don't seems consistent.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.1` |
| `@typescript-eslint/parser`        | `2.3.1` |
| `TypeScript`                       | `3.6.2` |
| `ESLint`                           | `6.4.0` |
| `node`                             | `10.16.3` |
| `npm`                              | `6.11.3` |

**Comment by bradzacher** [comment]
This is a bug because `new Promise()` is a `NewExpression`, not a `CallExpression`.
The rule is only coded to include `CallExpression`s as valid for `allowTypedFunctionExpressions`.
